sysrib/server.go

```diff
@@ -53,25 +60,57 @@ type Server struct {
 	// diff for sending to the dataplane for programming.
 	resolvedRoutes map[RouteKey]*ResolvedRoute
 
-	dataplane Dataplane
+	dataplane dataplaneAPI
 }
 
-type Dataplane interface {
+type dataplaneAPI interface {
 	ProgramRoute(*ResolvedRoute) error
 }
 
+// Dataplane is a wrapper around dpb.HALClient to enable testing before
+// resolved route translation.
+//
+// TODO(wenbli): This is a temporary workaround due to the instability of the
+// API. Once the dataplane API is stable, then we'll want to test at the API
+// layer instead.
+type Dataplane struct {
+	dpb.HALClient
+}
+
+// programRoute programs the route in the dataplane, returning an error on failure.
+func (d *Dataplane) ProgramRoute(r *ResolvedRoute) error {
+	log.V(1).Infof("sysrib: programming resolved route: %+v", r)
+	rr, err := resolvedRouteToRouteRequest(r)
+	if err != nil {
+		return err
+	}
+	_, err = d.InsertRoute(context.Background(), rr)
+	return err
+}
+
 // NewServer instantiates server to handle client queries.
-func NewServer(dataplane Dataplane) (*Server, error) {
+//
+// If dp is nil, then a connection attempt is made.
+func NewServer(dp dataplaneAPI) (*Server, error) {
 	rib, err := NewSysRIB(nil)
 	if err != nil {
 		return nil, err
 	}
 
+	if dp == nil {
+		opts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
```

Comment:
use local.NewCredentials()

Reply:
Is this the yet newer way of specifying an insecure connection?

Reply:
Chatted offline, thanks.

Reply:
local.NewCredentials() is a ServerOption, and not a DialOption, so I can't quite use it. Since the current code follows the recommendation I'll keep it as is: https://pkg.go.dev/google.golang.org/grpc#WithInsecure, thanks.

Reply:
it's fine, but `grpc.Dial("", grpc.WithTransportCredentials(local.NewCredentials()))`